src/persist-client/src/internal/state.rs

```diff
@@ -1318,12 +1337,64 @@ where
         });
         match existing_id {
-            Some((schema_id, _)) => {
-                // TODO: Validate that the decoded schemas still produce records
-                // of the recorded DataType, to detect shenanigans. Probably
-                // best to wait until we've turned on Schema2 in prod and thus
-                // committed to the current mappings.
-                Break(NoOpStateTransition(Some(*schema_id)))
+            Some((schema_id, encoded_schemas)) => {
+                let schema_id = *schema_id;
+                let new_k_datatype = mz_persist_types::columnar::data_type::<K>(key_schema)
+                    .expect("valid key schema");
+                let new_v_datatype = mz_persist_types::columnar::data_type::<V>(val_schema)
+                    .expect("valid val schema");
+
+                let new_k_encoded_datatype = encoded_data_type(&new_k_datatype);
+                let new_v_encoded_datatype = encoded_data_type(&new_v_datatype);
+
+                // Check if the generated Arrow DataTypes have changed.
+                if encoded_schemas.key_data_type != new_k_encoded_datatype
+                    || encoded_schemas.val_data_type != new_v_encoded_datatype
+                {
+                    let old_k_datatype =
+                        decode_data_type(Bytes::clone(&encoded_schemas.key_data_type))
+                            .expect("failed to roundtrip Arrow DataType");
+                    let old_v_datatype =
+                        decode_data_type(Bytes::clone(&encoded_schemas.val_data_type))
+                            .expect("failed to roundtrip Arrow DataType");
+
+                    let k_atleast_as_nullable =
+                        crate::schema::is_atleast_as_nullable(&old_k_datatype, &new_k_datatype);
+                    let v_atleast_as_nullable =
+                        crate::schema::is_atleast_as_nullable(&old_v_datatype, &new_v_datatype);
+
+                    // If the Arrow DataType for `k` or `v` has changed, but it's only become more
+                    // nullable, then we allow in-place re-writing of the schema.
+                    match (k_atleast_as_nullable, v_atleast_as_nullable) {
+                        // TODO(parkmycar): Remove this one-time migration after v0.123 ships.
+                        (Ok(()), Ok(())) => {
+                            let key = Bytes::clone(&encoded_schemas.key);
+                            let val = Bytes::clone(&encoded_schemas.val);
+                            self.schemas.insert(
+                                schema_id,
+                                EncodedSchemas {
+                                    key,
+                                    key_data_type: new_k_encoded_datatype,
+                                    val,
+                                    val_data_type: new_v_encoded_datatype,
+                                },
+                            );
```

Comment:
We should give ourselves some way to check if we're hitting this path later than we'd expect. (I think a log is fine if wiring up a metric is a hassle... shouldn't be too high of a multiplicity.)

Reply:
Wired up a metric which should be easier to observe! We theoretically could write a query over our logs, but wiring that up is maybe harder than adding a prom metric haha